core/commands/ls.go

```diff
@@ -60,158 +67,243 @@ The JSON output contains type information.
+				if dir == nil {
+					links = dagnode.Links()
+				} else {
+					links, err = dir.Links(req.Context)
+					if err != nil {
+						return err
+					}
+				}
+				outputLinks := make([]LsLink, len(links))
+				for j, link := range links {
+					lsLink, err := makeLsLink(req, dserv, resolve, link)
+					if err != nil {
+						return err
+					}
+					outputLinks[j] = *lsLink
+				}
+				output[i] = LsObject{
+					Hash:  paths[i],
+					Links: outputLinks,
+				}
+			}
+
+			return cmds.EmitOnce(res, &LsOutput{output})
+		}
+
 		for i, dagnode := range dagnodes {
 			dir, err := uio.NewDirectoryFromNode(ro, dagnode)
 			if err != nil && err != uio.ErrNotADir {
-				res.SetError(fmt.Errorf("the data in %s (at %q) is not a UnixFS directory: %s", dagnode.Cid(), paths[i], err), cmdkit.ErrNormal)
-				return
+				return fmt.Errorf("the data in %s (at %q) is not a UnixFS directory: %s", dagnode.Cid(), paths[i], err)
 			}
 
-			var links []*ipld.Link
+			var linkResults <-chan unixfs.LinkResult
 			if dir == nil {
-				links = dagnode.Links()
+				linkResults = makeDagNodeLinkResults(req, dagnode)
 			} else {
-				links, err = dir.Links(req.Context())
-				if err != nil {
-					res.SetError(err, cmdkit.ErrNormal)
-					return
-				}
+				linkResults = dir.EnumLinksAsync(req.Context)
 			}
 
-			output[i] = LsObject{
-				Hash:  paths[i],
-				Links: make([]LsLink, len(links)),
-			}
+			for linkResult := range linkResults {
+				output := make([]LsObject, len(req.Arguments))
 
-			for j, link := range links {
-				t := unixfspb.Data_DataType(-1)
-
-				switch link.Cid.Type() {
-				case cid.Raw:
-					// No need to check with raw leaves
-					t = unixfs.TFile
-				case cid.DagProtobuf:
-					linkNode, err := link.GetNode(req.Context(), dserv)
-					if err == ipld.ErrNotFound && !resolve {
-						// not an error
-						linkNode = nil
-					} else if err != nil {
-						res.SetError(err, cmdkit.ErrNormal)
-						return
+				for i, path := range paths {
+					output[i] = LsObject{
+						Hash:  path,
+						Links: nil,
 					}
+				}
+				outputLinks := make([]LsLink, 1)
 
-					if pn, ok := linkNode.(*merkledag.ProtoNode); ok {
-						d, err := unixfs.FSNodeFromBytes(pn.Data())
-						if err != nil {
-							res.SetError(err, cmdkit.ErrNormal)
-							return
-						}
-						t = d.Type()
-					}
+				if linkResult.Err != nil {
+					return linkResult.Err
+				}
+				link := linkResult.Link
+				lsLink, err := makeLsLink(req, dserv, resolve, link)
+				if err != nil {
+					return err
 				}
-				output[i].Links[j] = LsLink{
-					Name: link.Name,
-					Hash: link.Cid.String(),
-					Size: link.Size,
-					Type: t,
+				outputLinks[0] = *lsLink
+				output[i].Links = outputLinks
+				if err = res.Emit(&LsOutput{output}); err != nil {
+					return err
 				}
 			}
 		}
-
-		res.SetOutput(&LsOutput{output})
+		return nil
 	},
-	Marshalers: cmds.MarshalerMap{
-		cmds.Text: func(res cmds.Response) (io.Reader, error) {
+	PostRun: cmds.PostRunMap{
```

Comment:
see conversation above -- @kevina preferred -- and I think he was right, the output from the run be the same when streamed as it was when it's not, and that requires logic in the postRun to produce the right text output. however, I can add a cmd.MakeTypedEncoder and use re.Emit in the PostRun instead.

Reply:
I was following the structure of ipfs add here.

Reply:
I see, `ipfs add` is probably a bad example as it uses PostRun mainly for progress bar. It should be possible to do this using `Encoders`, which is the preferred way.

Reply:
The problem with using Encoders is that they don't support Stderr output.  This is when I generally switch to PostRun.  Here that is not the case so we should still use an Encoder.

Reply:
So the issue basically has to do with the output of the header/footer lines surrounding the directory objects, which requires either passing information about where exactly you are in a stream in the objects themselves, or using a post run.

I originally had this setup with additional flags on the LSObject so that it correct output could be generated with an encoder (I was using cmd.MakeEncoder but converting to cmd.MakeTypeEncoder would have been trivial):
https://github.com/ipfs/go-ipfs/pull/5611/commits/135d1304143a2adbdb79abd3185f970addab23c1

However @kevina and @schomatis felt it would be better to not add any additional flags to the LSObject and instead handle the output re-arrangement at the CLI level:
https://github.com/ipfs/go-ipfs/pull/5611#discussion_r230522138
https://github.com/ipfs/go-ipfs/pull/5611#discussion_r230539744

So this was the rewrite based on that feedback.

I don't actually have a particular dog in this fight, and am fine with either way, but I guess I do agree that chaining together the right header/footer output around directories is a CLI level concern, that is analogous the progress bar on add. I am fine doing it either way -- and perhaps there is a hybrid approach I'm not seeing.

Reply:
ok I will try to code this without using PostRun, but I'm not sure how to get the right output between directories on a stream without maintaining some kind of state between calls to the Encoder

Reply:
@hannahhoward I agree that you will need to maintain some sort of state, I am not sure if you can do this with Encoders, so in that case you might need to use PostRun.  I think using `PostRun` is better than the extra flags.

Using PostRun might break using the API when using the text encoder, see https://github.com/ipfs/go-ipfs/pull/5674#issuecomment-436777547.  So perhaps a hybrid solution like @Stebalien proposed may be in order, but really I am just guessing now.

@Stebalien thoughts.